codec-http2/src/main/java/io/netty/handler/codec/http2/StreamBufferingEncoder.java

```diff
@@ -149,13 +149,13 @@ public ChannelFuture writeHeaders(ChannelHandlerContext ctx, int streamId, Http2
         if (closed) {
             return promise.setFailure(new Http2ChannelClosedException());
         }
-        if (isExistingStream(streamId) || connection().goAwayReceived()) {
+        if (isExistingStream(streamId) || canCreateStream()) {
             return super.writeHeaders(ctx, streamId, headers, streamDependency, weight,
                     exclusive, padding, endOfStream, promise);
         }
-        if (canCreateStream()) {
-            return super.writeHeaders(ctx, streamId, headers, streamDependency, weight,
-                    exclusive, padding, endOfStream, promise);
+        if (connection().goAwayReceived()) {
+            promise.setFailure(new Http2Exception(Http2Error.NO_ERROR, "GOAWAY received"));
```

Comment:
Hmm... Looks like we need a `Http2GoAwayException` for the RpcProgress.REFUSED logic.
https://github.com/grpc/grpc-java/blob/528ef63c5803f9ead3668d17b3ba99a1c6817ad2/netty/src/main/java/io/grpc/netty/NettyClientHandler.java#L642

We can either save the error+debug data or maybe we can just use Http2Error.REFUSED_STREAM here. It looks like Http2GoAwayException is using STREAM_CLOSED today, which seems broken as well. That should be REFUSED_STREAM. I don't know if changing that helps us at all here.

Reply:
Save the goAwayException now.

Reply:
Saving exceptions produces really weird results, as the backtrace is misleading. It can also retain more memory than we'd like. Save the details, not the exception.